src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml

```diff
@@ -220,19 +305,95 @@ end = struct
         : Bignum_bigint.t limbs_type =
       map x (Common.cvar_field_to_bignum_bigint_as_prover (module Circuit))
 
+    let one (type field)
+        (module Circuit : Snark_intf.Run with type field = field) : field t =
+      of_bignum_bigint (module Circuit) Bignum_bigint.one
+
     let to_bignum_bigint_as_prover (type field)
         (module Circuit : Snark_intf.Run with type field = field) (x : field t)
         : Bignum_bigint.t =
       let l0, l1, l2 = to_bignum_bigint_limbs_as_prover (module Circuit) x in
       Bignum_bigint.(l0 + (Common.two_to_limb * l1) + (two_to_2limb * l2))
 
+    let to_string_as_prover (type field)
+        (module Circuit : Snark_intf.Run with type field = field) a : string =
+      sprintf "%s" @@ Bignum_bigint.to_string
+      @@ to_bignum_bigint_as_prover (module Circuit) a
+
+    let is_zero (type field)
+        (module Circuit : Snark_intf.Run with type field = field) (x : field t)
+        : Circuit.Boolean.var =
+      let open Circuit in
+      let x0, x1, x2 = to_limbs x in
+      let x0_is_zero = Field.(equal x0 zero) in
+      let x1_is_zero = Field.(equal x1 zero) in
+      let x2_is_zero = Field.(equal x2 zero) in
+      Boolean.(x0_is_zero && x1_is_zero && x2_is_zero)
+
+    let equal_as_prover (type field)
+        (module Circuit : Snark_intf.Run with type field = field)
+        (left : field t) (right : field t) : bool =
+      let open Circuit in
+      let left0, left1, left2 =
+        to_field_limbs_as_prover (module Circuit) left
+      in
+      let right0, right1, right2 =
+        to_field_limbs_as_prover (module Circuit) right
+      in
+      Field.Constant.(
+        equal left0 right0 && equal left1 right1 && equal left2 right2)
+
+    let assert_equal (type field)
+        (module Circuit : Snark_intf.Run with type field = field)
+        (left : field t) (right : field t) : unit =
+      let open Circuit in
+      let left0, left1, left2 = to_limbs left in
+      let right0, right1, right2 = to_limbs right in
+      Field.Assert.equal left0 right0 ;
+      Field.Assert.equal left1 right1 ;
+      Field.Assert.equal left2 right2
+
+    let checked_const_of_bignum_bigint (type field)
```

Comment:
or is this one the one that generates those constraints? is that the meaning of prefix `checked`?

Reply:
Yes.  I could not think of a better name.

Reply:
okay